// substance-test/src/views/test_suite_view.js
"use strict";

var _ = require("underscore");
var Application = require("substance-application");
var View = Application.View;
var $$ = Application.$$;
var TestView = require("./test_view");

var TestSuiteView = function(testSuite) {
  View.call(this);

  this.testSuite = testSuite;
  this.testViews = {};
};

TestSuiteView.Prototype = function() {

  this.show = function() {
    this.$el.show();
  };

  this.hide = function() {
    this.$el.hide();
  };

  this.render = function() {
    this.el.innerHTML = "";
    this.el.classList.add("test-suite");

    var titleEl = $$("h2.name", {
      text: this.testSuite.name
    });
    var testsEl = $$(".tests");

    for (var i = 0; i < this.testSuite.tests.length; i++) {
      var test = this.testSuite.tests[i];
      var testView = new TestView(test, this);
      this.testViews[test.name] = testView;
      testsEl.appendChild(testView.render().el);
    };

    this.el.appendChild(titleEl);
    this.el.appendChild(testsEl);

    return this;
  };

  this.dispose = function() {
    View.prototype.dispose.call(this);
    for (var i = 0; i < this.testViews.length; i++) {
      this.testViews[i].dispose();
    }
  };

  this.updateTestReport = function(test, report) {
    var testView = this.testViews[test.name];
    testView.updateReport(report);
  };

  this.visit = function(suite) {
    _.each(suite.tests, function(test) {

    });
  };
};
TestSuiteView.Prototype.prototype = View.prototype;
TestSuiteView.prototype = new TestSuiteView.Prototype();

module.exports = TestSuiteView;
