// substance-test/src/controllers/test_center_controller.js
"use strict";

var Controller = require("substance-application").Controller;
var Test = require("../test");
var TestCenter = require("../views/test_center");

var TestCenterController = function() {
  Controller.call(this);
};

TestCenterController.Prototype = function() {

  var __super__ = Controller.prototype;

  this.initialize = function(newState, cb) {
    this.testRunner = new Test.Runner();
    this.createView().render();

    cb(null);
  };

  this.transition = function(newState, cb) {
    if (newState.id === "suite") {
      var suiteId = newState.suiteId;
      var suite = this.testRunner.suites[suiteId];
      this.testRunner.runSuite(suite, function(err) {
        if (err) {
          console.error(err);
        }
      });
      this.view.updateTitle("Test-Suite:" + suiteId);
    }
    else if (newState.id === "full") {
      this.testRunner.runAll(function(err) {
        if (err) {
          console.error(err);
        }
      });
      this.view.updateTitle("Test-Center: Run All Test-Suites");
    }
    cb(null);
  };

  this.dispose = function() {
    __super__.dispose.call(this);
    if (this.view) {
      this.view.dispose();
      this.view = null;
    }
  };

  this.createView = function() {
    if (!this.view) this.view = new TestCenter(this.testRunner);
    return this.view;
  };

};

// Exports
// --------

TestCenterController.Prototype.prototype = Controller.prototype;
TestCenterController.prototype = new TestCenterController.Prototype();

// TODO: why is this necessary?
// _.extend(TestCenterController.prototype, util.Events);

module.exports = TestCenterController;
